R/replace.R

```diff
@@ -179,18 +179,50 @@ str_replace_na <- function(string, replacement = "NA") {
 
 str_transform <- function(string, pattern, replacement) {
   loc <- str_locate(string, pattern)
-  str_sub(string, loc, omit_na = TRUE) <- replacement(str_sub(string, loc))
+  new <- replacement(str_sub(string, loc))
+  str_sub(string, loc, omit_na = TRUE) <- new
   string
 }
-str_transform_all <- function(string, pattern, replacement) {
+
+str_transform_all <- function(string, pattern, replacement, error_call = caller_env()) {
   locs <- str_locate_all(string, pattern)
 
-  for (i in seq_along(string)) {
-    for (j in rev(seq_len(nrow(locs[[i]])))) {
-      loc <- locs[[i]]
-      str_sub(string[[i]], loc[j, 1], loc[j, 2]) <- replacement(str_sub(string[[i]], loc[j, 1], loc[j, 2]))
-    }
+  old <- str_sub_all(string, locs)
+  idx <- chop_index(old)
+
+  # unchop list into a vector, apply replacement(), and then rechop back into
+  # a list
+  old_flat <- vctrs::vec_unchop(old, idx)
```

Comment:
Depending on how much you value readability over performance, you don't actually need to pass `idx` here since you are unchopping in the order the input was provided

Reply:
In dev vctrs:

```R
xs <- list("a", c("b", "c"), c("d", "e", "f"))

x <- list_unchop(xs)
x <- toupper(x)
x
#> [1] "A" "B" "C" "D" "E" "F"

sizes <- list_sizes(xs)
vec_partition(x, sizes)
```